File: client-kotlin-spring-restclient/client/src/main/kotlin/app/epistola/client/validation/schema/ValidatingGenerationApi.kt
package app.epistola.client.validation.schema

import app.epistola.client.api.GenerationApi
import app.epistola.client.api.TemplatesApi
import app.epistola.client.model.GenerateBatchRequest
import app.epistola.client.model.GenerateDocumentRequest
import app.epistola.client.model.GenerationJobResponse
import org.springframework.http.ResponseEntity

/**
 * A wrapper around [GenerationApi] that validates request data against
 * the template's JSON Schema before sending to the server.
 *
 * For single-document requests, validation errors are thrown immediately.
 * For batch requests, all items are validated and errors are collected
 * into a single [TemplateDataValidationException].
 *
 * ```kotlin
 * val generationApi = GenerationApi(restClient)
 * val templatesApi = TemplatesApi(restClient)
 * val validating = ValidatingGenerationApi(generationApi, templatesApi)
 * // Validates data against the template's schema before calling the server:
 * validating.generateDocument("my-tenant", request)
 * ```
 */
class ValidatingGenerationApi(
    private val delegate: GenerationApi,
    templatesApi: TemplatesApi,
    cache: SchemaCache = TtlSchemaCache(),
) {
    private val validator = TemplateSchemaValidator(
        templatesApi = templatesApi,
        cache = cache,
    )

    fun generateDocument(
        tenantId: String,
        generateDocumentRequest: GenerateDocumentRequest,
    ): GenerationJobResponse {
        validator.validate(tenantId, generateDocumentRequest.templateId, generateDocumentRequest.data)
        return delegate.generateDocument(tenantId, generateDocumentRequest)
    }

    fun generateDocumentWithHttpInfo(
        tenantId: String,
        generateDocumentRequest: GenerateDocumentRequest,
    ): ResponseEntity<GenerationJobResponse> {
        validator.validate(tenantId, generateDocumentRequest.templateId, generateDocumentRequest.data)
        return delegate.generateDocumentWithHttpInfo(tenantId, generateDocumentRequest)
    }

    fun generateDocumentBatch(
        tenantId: String,
        generateBatchRequest: GenerateBatchRequest,
    ): GenerationJobResponse {
        validateBatch(tenantId, generateBatchRequest)
        return delegate.generateDocumentBatch(tenantId, generateBatchRequest)
    }

    fun generateDocumentBatchWithHttpInfo(
        tenantId: String,
        generateBatchRequest: GenerateBatchRequest,
    ): ResponseEntity<GenerationJobResponse> {
        validateBatch(tenantId, generateBatchRequest)
        return delegate.generateDocumentBatchWithHttpInfo(tenantId, generateBatchRequest)
    }

    private fun validateBatch(tenantId: String, request: GenerateBatchRequest) {
        val allErrors = mutableListOf<TemplateDataValidationException.ValidationError>()
        for ((index, item) in request.items.withIndex()) {
            try {
                validator.validate(tenantId, item.templateId, item.data)
            } catch (e: TemplateDataValidationException) {
                allErrors += e.errors.map { error ->
                    error.copy(path = "items[$index]${error.path}")
                }
            }
        }
        if (allErrors.isNotEmpty()) {
            throw TemplateDataValidationException(allErrors)
        }
    }
}

File: contracts/api/clients/dotnet-httpclient/src/Epistola.Client/Validation/Schema/ValidatingGenerationApi.cs
using System.Collections.Generic;
using System.Linq;
using Epistola.Client.Api;
using Epistola.Client.Client;
using Epistola.Client.Model;

namespace Epistola.Client.Validation.Schema;

/// <summary>
/// A wrapper around <see cref="IGenerationApi"/> that validates request data against the template's
/// JSON Schema before sending it to the server.
///
/// For single-document requests, validation errors are thrown immediately. For batch requests, all
/// items are validated and errors are collected into a single <see cref="TemplateDataValidationException"/>.
///
/// <code>
/// var validating = new ValidatingGenerationApi(generationApi, templatesApi);
/// validating.GenerateDocument("my-tenant", request);   // validates before calling the server
/// </code>
/// </summary>
public sealed class ValidatingGenerationApi
{
    private readonly IGenerationApi _delegate;
    private readonly TemplateSchemaValidator _validator;

    public ValidatingGenerationApi(IGenerationApi generationApi, ITemplatesApi templatesApi, ISchemaCache? cache = null)
    {
        _delegate = generationApi;
        _validator = new TemplateSchemaValidator(templatesApi, cache ?? new TtlSchemaCache());
    }

    public GenerationJobResponse GenerateDocument(string tenantId, GenerateDocumentRequest request)
    {
        _validator.Validate(tenantId, request.CatalogId, request.TemplateId, request.Data);
        return _delegate.GenerateDocument(tenantId, request);
    }

    public ApiResponse<GenerationJobResponse> GenerateDocumentWithHttpInfo(string tenantId, GenerateDocumentRequest request)
    {
        _validator.Validate(tenantId, request.CatalogId, request.TemplateId, request.Data);
        return _delegate.GenerateDocumentWithHttpInfo(tenantId, request);
    }

    public GenerationJobResponse GenerateDocumentBatch(string tenantId, GenerateBatchRequest request)
    {
        ValidateBatch(tenantId, request);
        return _delegate.GenerateDocumentBatch(tenantId, request);
    }

    public ApiResponse<GenerationJobResponse> GenerateDocumentBatchWithHttpInfo(string tenantId, GenerateBatchRequest request)
    {
        ValidateBatch(tenantId, request);
        return _delegate.GenerateDocumentBatchWithHttpInfo(tenantId, request);
    }

    private void ValidateBatch(string tenantId, GenerateBatchRequest request)
    {
        var allErrors = new List<TemplateDataValidationException.ValidationError>();
        for (var index = 0; index < request.Items.Count; index++)
        {
            var item = request.Items[index];
            try
            {
                _validator.Validate(tenantId, item.CatalogId, item.TemplateId, item.Data);
            }
            catch (TemplateDataValidationException e)
            {
                allErrors.AddRange(e.Errors.Select(error => error with { Path = $"items[{index}]{error.Path}" }));
            }
        }

        if (allErrors.Count > 0)
        {
            throw new TemplateDataValidationException(allErrors);
        }
    }
}
